compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirFunctionParameterChecker.kt

```diff
@@ -20,6 +24,25 @@ import org.jetbrains.kotlin.types.AbstractTypeChecker
 object FirFunctionParameterChecker : FirFunctionChecker() {
     override fun check(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
         checkVarargParameters(declaration, context, reporter)
+        checkParameterTypes(declaration, context, reporter)
+    }
+
+    private fun checkParameterTypes(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        for (valueParameter in declaration.valueParameters) {
+            val returnTypeRef = valueParameter.returnTypeRef
+            if (returnTypeRef !is FirErrorTypeRef) continue
+            // type problems on real source are already reported by ConeDiagnostic.toFirDiagnostics
+            if (returnTypeRef.source?.kind !is FirFakeSourceElementKind?) continue
+
+            val diagnostic = returnTypeRef.diagnostic
+            if (diagnostic !is ConeSimpleDiagnostic) continue
```

Comment:
Well, this is based on my suggestion: https://github.com/JetBrains/kotlin/pull/4203#discussion_r593638426 Previously, it looked like:
```kt
when (val diagnostic = returnTypeRef.diagnostic) {
     is ConeSimpleDiagnostic -> if (diagnostic.kind == DiagnosticKind.ValueParameterWithNoTypeAnnotation) reporter.reportOn(
                ...
```

Reply:
TBH I don't have opinion on this either way. I think this logic is simple enough that all the proposed ways are equally clear to me. I can flip and join the if here unless jsjeon thinks otherwise.

Reply:
We could mix. :) I still prefer the first half of my suggestion where other bail-out points were checked individually. It might be better to guard `reportOn` with a specific condition.

Reply:
Yeah that's fine. It's harder to reason out and keep track in your had what something is _not_ as opposed to what it _is, especially without a comment to explain where you are at. `if (returnTypeRef.source?.kind !is FirFakeSourceElementKind?) continue` (from previous revision) blew up my brain with so many negations. 🤯

Reply:
So it sounds like the current version is what we have all agreed on right?

Reply:
@tgeng Current version looks better to me, future readers will be happier 🙂